### PR TITLE
Add some defensive checks against invalid serialization

### PR DESCRIPTION
**Pull request by jrose-apple**
If a non-serializable Decl or Type makes it into serialization logic in a release build, an invalid swiftmodule might be produced silently, leading to crashes in clients. Try to catch this with an on-in-Release check that *something* was serialized. (This is my best guess at how rdar://problem/47459323 occurred.)

### CODE REVIEW

**Comment by jrose-apple** [comment]
@swift-ci Please test

File: lib/Serialization/Serialization.cpp [hunk 1]
```diff
@@ -3677,14 +3686,23 @@ static TypeAliasDecl *findTypeAliasForBuiltin(ASTContext &Ctx, Type T) {
...
   auto id = DeclAndTypeIDs[ty];
   assert(id != 0 && "type not referenced properly");
   (void)id;
 
   assert((id - 1) == TypeOffsets.size());
-
+  assert((DeclOffsets.empty() || DeclOffsets.back() != Out.GetCurrentBitNo()) &&
+         "encoding Decl and Type to the same offset");
   TypeOffsets.push_back(Out.GetCurrentBitNo());
+  SWIFT_DEFER {
```
Comment (harlanhaskins) [inline lib/Serialization/Serialization.cpp]: If this is the last statement in the function, why `SWIFT_DEFER` it?

Comment (jrose-apple) [reply lib/Serialization/Serialization.cpp]: It's…not?

Comment (harlanhaskins) [reply lib/Serialization/Serialization.cpp]: 🤦‍♂️ misread the braces, apologies.

**Review by harlanhaskins — APPROVED** [review]

**Review by slavapestov — APPROVED** [review]
Ah, makes sense.

**Pull request merged**